File: Lista de Exercício/Golang/Carro.go
// Questão 1, 2 e 6

package main

import (
	"fmt"
)

type Motor struct {
	Tipo     string
	Potencia int
}

func (m Motor) String() string {
	return fmt.Sprintf("%s - %d cv", m.Tipo, m.Potencia)
}

type Carro struct {
	Marca      string
	Modelo     string
	Ano        int
	Velocidade int
	Motor      Motor
}

func (c *Carro) NovoMotor(tipoMotor string) {
	fmt.Printf("Motor do %s: %s\n", c.Modelo, tipoMotor)
}

func (c *Carro) Bateria(capacidadeBateria int) {
	fmt.Printf("Bateria do %s: %d kWh\n", c.Modelo, capacidadeBateria)
}

func (c *Carro) Pneus(tipoPneu string) {
	fmt.Printf("Pneus do %s: %s\n", c.Modelo, tipoPneu)
}

func (c *Carro) Sobre() {
	fmt.Printf("%s %s (%d)\n", c.Marca, c.Modelo, c.Ano)
	fmt.Printf("Motor: %s\n", c.Motor)
}

func (c *Carro) Acelerar(incremento int) {
	c.Velocidade += incremento
	fmt.Printf("%s acelerou para %d km/h\n", c.Modelo, c.Velocidade)
}

func (c *Carro) Frear(decremento int) {
	if c.Velocidade-decremento < 0 {
		c.Velocidade = 0
	} else {
		c.Velocidade -= decremento
	}
	fmt.Printf("%s freou para %d km/h\n", c.Modelo, c.Velocidade)
}

func (c *Carro) ExibirVelocidade() {
	fmt.Printf("Velocidade do %s: %d km/h\n", c.Modelo, c.Velocidade)
}

func imprimirSeparador() {
	fmt.Println("========================================")
}

func main() {
	carro1 := &Carro{Marca: "Toyota", Modelo: "Corolla", Ano: 2020, Motor: Motor{Tipo: "V6", Potencia: 200}}
	carro2 := &Carro{Marca: "Honda", Modelo: "Civic", Ano: 2018, Motor: Motor{Tipo: "V4", Potencia: 180}}
	carro3 := &Carro{Marca: "Ford", Modelo: "Mustang", Ano: 2022, Motor: Motor{Tipo: "V8", Potencia: 450}}

	carros := []*Carro{carro1, carro2, carro3}

	for _, carro := range carros {
		imprimirSeparador()
		carro.Sobre()
		if carro.Marca == "Toyota" {
			carro.Bateria(12)
			carro.Pneus("Radiais")
		} else if carro.Marca == "Honda" {
			carro.Bateria(14)
			carro.Pneus("Esportivos")
		} else {
			carro.Bateria(15)
			carro.Pneus("Off-road")
		}
	}

	imprimirSeparador()
	fmt.Println("\nTeste de aceleração e frenagem (Toyota Corolla):")
	imprimirSeparador()

	carro1.Acelerar(50)
	carro1.ExibirVelocidade()
	carro1.Acelerar(30)
	carro1.ExibirVelocidade()
	carro1.Frear(20)
	carro1.ExibirVelocidade()
	carro1.Frear(100)
	carro1.ExibirVelocidade()

	imprimirSeparador()
}

File: Lista de Exercício/Java/Carro.java
// Questões 1, 2 e 6

package ListaDeExercicio.Java;

public class Carro {

    private String marca;
    private String modelo;
    private int ano;
    private int velocidade;
    private Motor motor;

    public Carro(String marca, String modelo, int ano, String tipoMotor, int potenciaMotor) {
        this.marca = marca;
        this.modelo = modelo;
        this.ano = ano;
        this.velocidade = 0;
        this.motor = new Motor(tipoMotor, potenciaMotor);
    }

    public String getMarca() {
        return marca;
    }

    public void mostrarMotor() {
        System.out.println("Motor do " + modelo + ": " + motor.toString());
    }

    public void mostrarBateria(int capacidadeBateria) {
        System.out.println("Bateria do " + modelo + ": " + capacidadeBateria + " kWh");
    }

    public void mostrarPneus(String tipoPneu) {
        System.out.println("Pneus do " + modelo + ": " + tipoPneu);
    }

    public void sobre() {
        System.out.println("Carro: " + marca + " " + modelo + " (" + ano + ")");
        System.out.println("Motor: " + motor.toString());
    }

    public void acelerar(int incremento) {
        velocidade += incremento;
        System.out.println(modelo + " acelerou para " + velocidade + " km/h");
    }

    public void frear(int decremento) {
        velocidade = Math.max(0, velocidade - decremento);
        System.out.println(modelo + " freou para " + velocidade + " km/h");
    }

    public void exibirVelocidade() {
        System.out.println("Velocidade do " + modelo + ": " + velocidade + " km/h");
    }

    public static void imprimirSeparador() {
        System.out.println("========================================");
    }

    public static void main(String[] args) {
        // Criando objetos Carro
        Carro carro1 = new Carro("Toyota", "Corolla", 2020, "V6", 200);
        Carro carro2 = new Carro("Honda", "Civic", 2018, "V4", 180);
        Carro carro3 = new Carro("Ford", "Mustang", 2022, "V8", 450);

        Carro[] carros = {carro1, carro2, carro3};

        for (Carro carro : carros) {
            imprimirSeparador();
            carro.sobre();
            int capacidadeBateria = carro.getMarca().equals("Toyota") ? 12 : carro.getMarca().equals("Honda") ? 14 : 15;
            carro.mostrarBateria(capacidadeBateria);
            String tipoPneu = carro.getMarca().equals("Toyota") ? "Radiais" : carro.getMarca().equals("Honda") ? "Esportivos" : "Off-road";
            carro.mostrarPneus(tipoPneu);
        }

        imprimirSeparador();
        System.out.println("\nTeste de aceleração e frenagem (Toyota Corolla):");
        imprimirSeparador();

        carro1.acelerar(50);
        carro1.exibirVelocidade();
        carro1.acelerar(30);
        carro1.exibirVelocidade();
        carro1.frear(20);
        carro1.exibirVelocidade();
        carro1.frear(100);
        carro1.exibirVelocidade();

        imprimirSeparador();
    }
}

class Motor {
    private String tipo;
    private int potencia;

    public Motor(String tipo, int potencia) {
        this.tipo = tipo;
        this.potencia = potencia;
    }

    @Override
    public String toString() {
        return tipo + " - " + potencia + " cv";
    }
}
